### PR TITLE
[#11572] Notification feature - User frontend tests

### PR DESCRIPTION
**Pull request by fsgmhoward**
Part of #11572

This PR contains a number of frontend tests for the user side of the notification feature. This includes the banner and the notification page.

Will wait for the E2E PR to merge first before this one as that comes with potential breaking changes.

### CODE REVIEW

**Review by zhaojj2209 — APPROVED** [review]
LGTM, good coverage!

File: src/web/app/components/notification-banner/notification-banner.component.spec.ts [hunk 1]
```diff
@@ -23,11 +59,90 @@ describe('NotificationBannerComponent', () => {
...
+
+  it('should close after clicking mark as read', () => {
+    jest.spyOn(notificationService, 'getUnreadNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [testNotificationOne],
+    }));
+    jest.spyOn(notificationService, 'markNotificationAsRead').mockReturnValue(of({
+      readNotifications: [testNotificationOne.notificationId],
+    }));
+    const spy1: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
+    .mockImplementation((request: MarkNotificationAsReadRequest) => {
```
Comment (jianhandev) [inline src/web/app/components/notification-banner/notification-banner.component.spec.ts:94]: I think there's an indentation here

File: src/web/app/components/notification-banner/notification-banner.component.spec.ts [hunk 2]
```diff
@@ -23,11 +59,90 @@ describe('NotificationBannerComponent', () => {
...
+      expect(request.notificationId).toEqual(testNotificationOne.notificationId);
+      return of({
+        readNotifications: [request.notificationId],
+      });
+    });
+    const spy2: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
```
Comment (jianhandev) [inline src/web/app/components/notification-banner/notification-banner.component.spec.ts:100]:
```suggestion
    const messageSpy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
```
Try to use more descriptive names

File: src/web/app/components/notification-banner/notification-banner.component.spec.ts [hunk 3]
```diff
@@ -23,11 +59,90 @@ describe('NotificationBannerComponent', () => {
...
+  });
+
+  it('should close after clicking mark as read', () => {
+    jest.spyOn(notificationService, 'getUnreadNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [testNotificationOne],
+    }));
+    jest.spyOn(notificationService, 'markNotificationAsRead').mockReturnValue(of({
+      readNotifications: [testNotificationOne.notificationId],
+    }));
+    const spy1: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
```
Comment (jianhandev) [inline src/web/app/components/notification-banner/notification-banner.component.spec.ts:93]:
```suggestion
    const notificationSpy: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
```

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 4]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+    fixture.detectChanges();
+    expect(component.notificationTabs[0].hasTabExpanded).toBeTruthy();
+    fixture.debugElement.query(By.css('#btn-mark-as-read')).nativeElement.click();
+    expect(spy1).toHaveBeenCalledTimes(1);
+    expect(spy2).toHaveBeenCalledTimes(1);
+    // check that it is no longer expanded
+    expect(component.notificationTabs[0].hasTabExpanded).toBeFalsy();
+  });
+
+  it('should toggle the tab when the header is clicked', () => {
```
Comment (jianhandev) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:105]:
```suggestion
  it('should collapse an expanded tab when the header is clicked', () => {
```

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 5]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+  it('should mark notification as read when button is clicked', () => {
+    jest.spyOn(notificationService, 'getAllNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [testNotificationOne],
+    }));
+    jest.spyOn(notificationService, 'getReadNotifications').mockReturnValue(of({
+      readNotifications: [],
+    }));
+    const spy1: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
+    .mockImplementation((request: MarkNotificationAsReadRequest) => {
+      expect(request.notificationId).toEqual(testNotificationOne.notificationId);
```
Comment (jianhandev) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:85]: indentation

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 6]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+      return of({
+        readNotifications: [request.notificationId],
+      });
+    });
+    const spy2: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
+    .mockImplementation((args: string) => {
+      expect(args).toEqual('Notification marked as read.');
```
Comment (jianhandev) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:92]: indentation

**Review by jianhandev — COMMENTED** [review]
Looks good, mostly indentation issues and naming.

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts (hunk 5 above)
Comment (fsgmhoward) [reply src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:85]: Thanks for the comment. I will fix all in the next commit (the same indentation will be updated for the admin side test).

**Review by jianhandev — APPROVED** [review]
LGTM

File: src/web/app/components/notification-banner/notification-banner.component.spec.ts [hunk 7]
```diff
@@ -23,11 +59,90 @@ describe('NotificationBannerComponent', () => {
...
+  });
+
+  it('should snap with no unread notifications', () => {
+    jest.spyOn(notificationService, 'getUnreadNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [],
+    }));
+    component.ngOnInit();
+    expect(component.notifications.length).toEqual(0);
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
```
Comment (wkurniawan07) [inline src/web/app/components/notification-banner/notification-banner.component.spec.ts:126]: This is not the ideal way of snapshot testing. Here what you're doing is "have a starting state, do something, then the snapshot should be like this". What should be done is "have a state, then the snapshot should be like this".

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 8]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+    component.isLoadingNotifications = true;
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
+  });
+
+  it('should snap when it fails to load', () => {
+    component.isLoadingNotifications = false;
+    component.hasLoadingFailed = true;
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
```
Comment (wkurniawan07) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:127]: Contrast to this. This is ideal.

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 9]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+  });
+
+  it('should snap when it loads the provided notifications', () => {
+    jest.spyOn(notificationService, 'getAllNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [testNotificationOne, testNotificationTwo],
+    }));
+    component.loadNotifications();
+    expect(component.notificationTabs.length).toEqual(2);
+    fixture.detectChanges();
+    expect(fixture).toMatchSnapshot();
```
Comment (wkurniawan07) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:143]: This is not ideal either.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Need a better way to create the snapshot test cases

**Review by ziqing26 — APPROVED** [review]
LGTM overall!

File: src/web/app/components/notification-banner/notification-banner.component.spec.ts (hunk 7 above)
Comment (fsgmhoward) [reply src/web/app/components/notification-banner/notification-banner.component.spec.ts:126]: Thank you for the comments. I will try to improve it in the following commits.

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 10]
```diff
@@ -26,11 +63,126 @@ describe('UserNotificationsListComponent', () => {
...
+  });
+
+  it('should snap when all loaded notifications are read', () => {
+    jest.spyOn(notificationService, 'getAllNotificationsForTargetUser').mockReturnValue(of({
+      notifications: [testNotificationOne, testNotificationTwo],
+    }));
```
[the remark this answers is not in the document]
Comment (fsgmhoward) [reply src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:149]: Currently the `spyOn` has different set of values for different test cases. However, I guess I can remove it from the snapshot tests. I will see how it goes after implementing the improvement from Wilson's comment. Thanks for pointing this out!

**Review by wkurniawan07 — COMMENTED** [review]
I'm ok to accept the PR as is, but some parts can be simplified

File: src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts [hunk 11]
```diff
@@ -26,11 +82,134 @@ describe('UserNotificationsListComponent', () => {
...
+    const getReadNotificationSpy = jest.spyOn(notificationService, 'getReadNotifications').mockReturnValue(of({
+      readNotifications: [
+        testNotificationOne.notificationId,
+      ],
+    }));
+    component.loadNotifications();
+    expect(getNotificationSpy).toHaveBeenCalledTimes(1);
+    expect(getReadNotificationSpy).toHaveBeenCalledTimes(1);
+    expect(component.notificationTabs.length).toEqual(2);
```
[the remark this answers is not in the document]
Comment (fsgmhoward) [reply src/web/app/components/user-notifications-list/user-notifications-list.component.spec.ts:112]: Noted. I will update it in the following commit. Thanks for pointing it out!

**Review by wkurniawan07 — APPROVED** [review]

**Pull request merged**